### PR TITLE
G4.49 Implement a GermplasmNameExists validator

### PR DESCRIPTION
**Pull request by carolyncaron**
**Issue #49**

## Motivation
We need a validator which checks that a germplasm name already exists in the database, in the case that we are importing data related to that germplasm (such as phenotypes, genotypes, other germplasm). Since germplasm names are only required to be unique within a single genus, we also need a Trait that can set and get the genus for this validator.

## What does this PR do?
1) Creates an `Organism` trait that includes 2 setters and 1 getter:
   - `setOrganismID()` for when validation should be limited to a specific organism
   - `setGenus()` for when validation should extend to an entire genus
   - `getOrganismIDs()` for retrieving a set of organism IDs. Depending on the setter used, this will be 1 organism ID, or all organism IDs for a genus.
   
2) Creates a `GermplasmNameExists` validator. 
   - This is a row-level validator
   - Uses the `Organism` and `ColumnIndices` traits
   - Failed cases include:
      - When a germplasm name is missing in the database
      - When a germplasm name occurs in the database more than once
      - When both of the above cases occur within the same input row

3) Implements tests for both 1) and 2).

### Automated Testing
**Organism Trait**
- Fake Validators
   - `ValidatorOrganism.php`
   - `ValidatorOrganismNOConnection.php`
- Traits
   - `ValidatorTraitOrganismTest.php` -> `testOrganismSetterGetter()`
   - `ValidatorTraitOrganismMissingConnectionTest.php` -> `testOrganismSetters()`

**GermplasmNameExists Validator**
- `ValidatorGermplasmNameExistsTest.php` -> `testValidatorGermplasmNameExists()`
   - Data Provider: `provideRowToGermplasmNameExists()`

### Manual Testing

1. Start a fresh docker image/container on this branch. **You can use the devcontainer approach OR the following commands:**
  ```
  cd ~/Dockers
  git clone https://github.com/TripalCultivate/TripalCultivate trpcultBase-50
  cd trpcultBase-50
  git checkout g4.49-germplasmLookupValidator
  docker build --tag=trpcultivate-base:review50 ./
  docker run --publish=80:80 -tid --name=base50 --volume=`pwd`:/var/www/drupal/web/modules/contrib/TripalCultivate trpcultivate-base:review50
  ```

2. Login to your TripalCultivate site and navigate to `PHP Execute` (you may need to enable it in your devel dropdown first). Create an instance of the validator and enter organism(s) and germplasm name(s) and finally, your row values:
```
$validator = \Drupal::service('plugin.manager.trpcultivate_validator')
  ->createInstance('germplasm_name_exists');

$chado_connection = \Drupal::service('tripal_chado.database');

// Insert organism(s) into chado.
$genus = 'Tripalus';
$species = 'databasica';
$organism_id = $chado_connection->insert('1:organism')
      ->fields([
        'genus' => $genus,
        'species' => $species,
      ])
      ->execute();

// Insert germplasm name(s) into chado.
$germplasm_name = 'TESTNAME';
$stock_values = [
      'organism_id' => $organism_id,
      'name' => $germplasm_name,
      'uniquename' => 'TEST:1',
      'type_id' => 9,
];
$stock_id = $chado_connection->insert('1:stock')
   ->fields($stock_values)->execute();

// To insert a duplicate, insert again with the same 'name' field.

// Test row values
$row_values = [
   1 => 'Column1',
   2 => $germplasm_name,
   3 => 'Column3',
];

// Set indices to the columns you want the validator to check.
$indices = [2];

$validator->setIndices($indices);
$validator->setOrganismID($organism_id);
$validation_status = $validator->validateRow($row_values);
dpm($validation_status);
```

### CODE REVIEW

**Comment by reynoldtan** [comment]
Impressive job @carolyncaron. I performed manual test and tested all 3 failed items mentioned - duplicate, missing and both in a data row and all worked seamlessly.

I have comments about Organism trait.

- [ ] I would include the operator '=' (3rd parameter) even though it is optional  https://github.com/TripalCultivate/TripalCultivate/blob/3513cd5505857c074f45cea2d9134f42e194ee43/src/TripalCultivateValidator/ValidatorTraits/Organism.php#L38
- [ ] I wonder if the db check is necessary. In the context of the importer where it extends to ChadoImporterBase, I think this is already verfied. https://github.com/TripalCultivate/TripalCultivate/blob/3513cd5505857c074f45cea2d9134f42e194ee43/src/TripalCultivateValidator/ValidatorTraits/Organism.php#L32
- [ ] The name of setter and getter is causing a Problem of
image

File: src/Plugin/Validators/GermplasmNameExists.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+   * @param array $row_values
+   *   An array of values from a single row/line in the file where each key maps
+   *   to a column index and each value is the content of that column.
+   *
+   * @return array
+   *   An associative array with the following keys.
+   *   - 'case': a developer-focused string describing the case checked.
+   *   - 'valid': TRUE if the germplasm exists in the database, FALSE otherwise.
+   *   - 'failedItems': an array of items that failed with the following keys.
+   *     This is an empty array if the data row input was valid.
```
Comment (laceysanderson) [inline src/Plugin/Validators/GermplasmNameExists.php:87]: This should document the structure of the failedItems array. Right now it says "an array of items that failed with the following keys." but does not elaborate what those keys are.

File: src/Plugin/Validators/GermplasmNameExists.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    $duplicate = FALSE;
+    $failedItems = [];
+
+    // Iterate through our array of row values.
+    foreach ($row_values as $index => $cell) {
+      // Only validate the values in which their index is also within our
+      // context array of indices.
+      if (in_array($index, $indices)) {
+        // Trim the contents of our cell in case we have flanking whitespace.
+        $cell = trim($cell);
```
Comment (laceysanderson) [inline src/Plugin/Validators/GermplasmNameExists.php:136]: What about the case where the cell is empty? Currently it looks like the database will be queried for a stock with an empty name (which may be present as the column is not required) but it would then the validator would pass if it found one, right? That seems spurious. Rather it might be better to only do the query if the cell is not empty.

Now the question is: what should be the validation status if the cell is empty? Does this column also act as a "Required Cell" validator in which case it should fail for empty cell...

File: src/TripalCultivateValidator/ValidatorTraits/Organism.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      ->condition('o.genus', $genus);
+    $record = $query->execute()->fetchCol();
+    // If we have 1+ organisms with the genus, return as an array.
+    if (count($record) > 0) {
+      $this->context['organism_ids'] = $record;
+    }
+    else {
+      // Since this is a user-provided value, the error is going to be logged
+      // instead of thrown as an exception and then checked by a validator so
+      // that the error can be passed to the user in a friendly way.
```
Comment (laceysanderson) [inline src/TripalCultivateValidator/ValidatorTraits/Organism.php:80]: This comment led me down a bit of a rabbit hole because I read this to mean that the GermplasmNameExists validator would provide this feedback. However it can't because the getter throws an exception since the context key is not set in this case. We discussed this and I'll create an issue about this as the approach here is consistent with what was done in the GenusConfigured trait.

**Review by laceysanderson — COMMENTED** [review]
I've made it through the organism trait and the germplasmNameExists validator. I will come back to review the tests tomorrow.

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismMissingConnectionTest.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    }
+    catch (\Exception $e) {
+      $exception_caught = TRUE;
+      $exception_message = $e->getMessage();
+    }
+    $this->assertTrue(
+      $exception_caught,
+      "Calling setOrganismID() when missing an instance of ChadoConnection should throw an exception, but it did not."
+    );
+    $this->assertStringContainsString(
```
Comment (laceysanderson) [inline tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismMissingConnectionTest.php]: Does this need to be stringContainsString()? I think the expected message is the full exception message so perhaps we should use equals here?

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismMissingConnectionTest.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    }
+    catch (\Exception $e) {
+      $exception_caught = TRUE;
+      $exception_message = $e->getMessage();
+    }
+    $this->assertTrue(
+      $exception_caught,
+      "Calling setGenus() when missing an instance of ChadoConnection should throw an exception, but it did not."
+    );
+    $this->assertStringContainsString(
```
Comment (laceysanderson) [inline tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismMissingConnectionTest.php]: Same thing as above, perhaps we should use assertEquals() here?

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    $first_organism_id = $this->chado_connection->insert('1:organism')
+      ->fields([
+        'genus' => $genus,
+        'species' => $species,
+      ])
+      ->execute();
+    $this->assertIsNumeric($first_organism_id, 'We were not able to create the organism ' . $genus . ' ' . $species . ' in Chado for testing.');
+    // Cast our ID to an int since querying Chado gives us a string.
+    $first_organism_id = (int) $first_organism_id;
+    $this->instance->setOrganismID($first_organism_id);
```
Comment (laceysanderson) [inline tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php:165]: It would be good to check there was not a logged message here.

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+      $grabbed_organism_ids,
+      "We expected getOrganismIDs() to return an array, but it did not."
+    );
+    // Check that we were returned an array with 2 separate IDs.
+    $this->assertCount(2, $grabbed_organism_ids, "We expected getOrganismIDs to return an array with 2 IDs, but it contained a different amount.");
+    $this->assertContains(
+      $second_organism_id,
+      $grabbed_organism_ids,
+      "The organism ID retrieved using getOrganismIDs() is not the same as the ID given to setOrganismID()."
+    );
```
Comment (laceysanderson) [inline tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php]: Might it be better to check here that both the ids returned are the expected ones? Rather then only checking the second one?

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    $second_organism_id = $this->chado_connection->insert('1:organism')
+      ->fields([
+        'genus' => $genus,
+        'species' => $species,
+      ])
+      ->execute();
+    $this->assertIsNumeric($second_organism_id, 'We were not able to create the organism ' . $genus . ' ' . $species . ' in Chado for testing.');
+
+    // Set the genus to 'Tripalus'.
+    $this->instance->setGenus($genus);
```
Comment (laceysanderson) [inline tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php:213]: Maybe you should check that no errors were logged here since we don't expect any?

File: tests/src/Kernel/Validators/ValidatorGermplasmNameExistsTest.php [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+   *       - 'duplicate_cells' (OPTIONAL): Present if a germplasm name is
+   *         duplicated in the database.
+   *         - 1+ arrays keyed by the column number (first column = 1) in the
+   *          input row that contains a duplicate germplasm name, further keyed
+   *          by:
+   *          - 'germplasm_name': The name of the duplicate germplasm.
+   *          - 'duplicates': A list of indices in $duplicate_insert_values
+   *            which correspond to records that are expected to be duplicated.
+   */
+  public function provideRowToGermplasmNameExists() {
```
Comment (laceysanderson) [inline tests/src/Kernel/Validators/ValidatorGermplasmNameExistsTest.php:188]: It would be good to add a test scenario here where the germplasm name is empty. That said, we still need to decide what we expect in such a scenario :see_no_evil:

**Review by laceysanderson — COMMENTED** [review]
Small suggestions in the tests but overall these look good.

File: src/Plugin/Validators/GermplasmNameExists.php (hunk 2 above)
Comment (laceysanderson) [reply src/Plugin/Validators/GermplasmNameExists.php:136]: After discussion, we've decided to create an issue for a more ideal handling of this BUT in the meantime, we will check for empty cell here and validation will fail with it's own case message indicating we require a value in this cell.

**Comment by carolyncaron** [comment]
Replying to @reynoldtan's review:

> * [x]  I would include the operator '=' (3rd parameter) even though it is optional  https://github.com/TripalCultivate/TripalCultivate/blob/3513cd5505857c074f45cea2d9134f42e194ee43/src/TripalCultivateValidator/ValidatorTraits/Organism.php#L38

Thanks! This is addressed in commit 6c25a1709cf0aab9bf0b10a927775a50b062157b now 😄 

> * [x]  I wonder if the db check is necessary. In the context of the importer where it extends to ChadoImporterBase, I think this is already verfied. https://github.com/TripalCultivate/TripalCultivate/blob/3513cd5505857c074f45cea2d9134f42e194ee43/src/TripalCultivateValidator/ValidatorTraits/Organism.php#L32

I used the GenusConfigured trait as a template for Organism, and it includes this check, so I'd like to keep it for consistency. I actually found it very helpful while I was implementing the tests and then the GermplasmNameExists validator, since it would inform me if I had done it incorrectly 😅 

> * [x]  The name of setter and getter is causing a Problem of
>   image

We had a brief chat about this, but we ultimately decided as a team to keep the names as-is instead of resorting to "Id" to comply with camelcase standards.

File: src/Plugin/Validators/GermplasmNameExists.php (hunk 1 above)
Comment (carolyncaron) [reply src/Plugin/Validators/GermplasmNameExists.php:87]: Apparently I focused on documenting the tests and forgot the validator itself 😆 Addressed in commit 4136623035a74de235583fcfee1eec6ec66a9927!

File: src/Plugin/Validators/GermplasmNameExists.php (hunk 2 above)
Comment (carolyncaron) [reply src/Plugin/Validators/GermplasmNameExists.php:136]: Addressed in commits 42bb27707600cc5ff88d5d284e7e7996a5b9f0e2 and 11d58bf1b76e56e1195fff0ffb4e529c0ef8a004 😊

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismMissingConnectionTest.php (hunk 4 above)
Comment (carolyncaron) [reply tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismMissingConnectionTest.php]: Good catch! Addressed in commit c811c9524d36cc7a25c007285bb46d2044b79af2 (there were also some in `ValidatorTraitOrganismTest`)

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismMissingConnectionTest.php (hunk 5 above)
Comment (carolyncaron) [reply tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismMissingConnectionTest.php]: Addressed in commit c811c9524d36cc7a25c007285bb46d2044b79af2

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php (hunk 6 above)
Comment (carolyncaron) [reply tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php:165]: Addressed in commit 3677eee80a83fb873d553763b67b29e8a8c7d270

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php (hunk 7 above)
Comment (carolyncaron) [reply tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php]: Addressed in commit 3677eee80a83fb873d553763b67b29e8a8c7d270

File: tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php (hunk 8 above)
Comment (carolyncaron) [reply tests/src/Kernel/Validators/Traits/ValidatorTraitOrganismTest.php:213]: Addressed in commit 3677eee80a83fb873d553763b67b29e8a8c7d270

File: tests/src/Kernel/Validators/ValidatorGermplasmNameExistsTest.php (hunk 9 above)
Comment (carolyncaron) [reply tests/src/Kernel/Validators/ValidatorGermplasmNameExistsTest.php:188]: Addressed in commit 42bb27707600cc5ff88d5d284e7e7996a5b9f0e2

**Review by reynoldtan — APPROVED** [review]

File: src/TripalCultivateValidator/ValidatorTraits/Organism.php (hunk 3 above)
Comment (laceysanderson) [reply src/TripalCultivateValidator/ValidatorTraits/Organism.php:80]: Concerns moved to issue #57

**Review by laceysanderson — APPROVED** [review]
All of my concerns are addressed! This is ready to merge 😍
